packages/transaction-controller/src/helpers/MultichainHelper.ts

```diff
@@ -0,0 +1,401 @@
+ * @property queryEntireHistory - Whether to initially query the entire transaction history or only recent blocks.
+ * @property updateTransactions - Whether to update local transactions using remote transaction data.
+ */
+type IncomingTransactionOptions = {
+  includeTokenTransfers?: boolean;
+  isEnabled?: () => boolean;
+  queryEntireHistory?: boolean;
+  updateTransactions?: boolean;
+};
+
+type Trackers = {
+  nonceTracker: NonceTracker;
+  pendingTransactionTracker: PendingTransactionTracker;
+  incomingTransactionHelper: IncomingTransactionHelper;
+};
+
+export class MultichainHelper {
+  #isMultichainEnabled: boolean;
+
+  readonly #ethQuery: EthQuery;
+
+  readonly #nonceTracker: NonceTracker;
+
+  readonly #incomingTransactionOptions: IncomingTransactionOptions;
+
+  readonly #findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+
+  readonly #getNetworkClientById: NetworkController['getNetworkClientById'];
+
+  readonly #getNetworkClientRegistry: NetworkController['getNetworkClientRegistry'];
+
+  readonly #removeIncomingTransactionHelperListeners: (
+    IncomingTransactionHelper: IncomingTransactionHelper,
+  ) => void;
+
+  readonly #removePendingTransactionTrackerListeners: (
+    pendingTransactionTracker: PendingTransactionTracker,
+  ) => void;
+
+  readonly #createTrackersForNetworkClient: (
+    networkClient: NetworkClient,
+    etherscanRemoteTransactionSource: EtherscanRemoteTransactionSource,
+  ) => Trackers;
+
+  readonly #acquireNonceLockForChainIdKey: (opts: {
+    chainId: Hex;
+    key?: string;
+  }) => Promise<() => void>;
+
+  readonly #trackingMap: Map<NetworkClientId, Trackers> = new Map();
+
+  readonly etherscanRemoteTransactionSourcesMap: Map<
+    Hex,
+    EtherscanRemoteTransactionSource
+  > = new Map();
+
+  constructor({
+    isMultichainEnabled,
+    ethQuery,
+    nonceTracker,
+    incomingTransactionOptions,
+    findNetworkClientIdByChainId,
+    getNetworkClientById,
+    getNetworkClientRegistry,
+    removeIncomingTransactionHelperListeners,
+    removePendingTransactionTrackerListeners,
+    createTrackersForNetworkClient,
+    acquireNonceLockForChainIdKey,
+  }: {
+    isMultichainEnabled: boolean;
+    ethQuery: EthQuery;
+    nonceTracker: NonceTracker;
+    incomingTransactionOptions: IncomingTransactionOptions;
+
+    findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+    getNetworkClientById: NetworkController['getNetworkClientById'];
+    getNetworkClientRegistry: NetworkController['getNetworkClientRegistry'];
+
+    removeIncomingTransactionHelperListeners: (
+      IncomingTransactionHelper: IncomingTransactionHelper,
+    ) => void;
+    removePendingTransactionTrackerListeners: (
+      pendingTransactionTracker: PendingTransactionTracker,
+    ) => void;
+    createTrackersForNetworkClient: (
+      networkClient: NetworkClient,
+      etherscanRemoteTransactionSource: EtherscanRemoteTransactionSource,
+    ) => Trackers;
```

Comment:
i found it simpler to instantiate the trackers in the TransactionController because there are several dependencies that would need to be passed into MultichainHelper otherwise

Reply:
I think instantiation makes total sense in the `TransactionController` given the many arguments it already knows, and the addition of the listeners, but we need to avoid delegating too much responsibility back.

Would a cleaner split be a callback per helper such as `createIncomingTransactionHelper` and `createPendingTransactionTracker` so the `TransactionController` knows as little as possible and isn't coupled to the concept of network client?

Reply:
done here https://github.com/MetaMask/core/pull/3896/commits/de7b5ac37b0842446ee8689acaec5b77dfee482e